File: solutions/lua/run-length-encoding/run-length-encoding.lua
local function encode_chunk(num, char)
  return num == 1 and char or num .. char
end

return {
  -- Generates a string where consecutive elements are a count and data value.
  encode = function(s)
    if s == '' then return '' end
    local out = ''

    local prev = s:sub(1, 1)
    local count = 1
    for n=2, s:len() do
      if s:sub(n,n) == prev then
        count = count + 1
      else
        out = out .. encode_chunk(count, prev)
        prev = s:sub(n, n)
        count = 1
      end
    end

    return out .. encode_chunk(count, prev)
  end,

  -- It should also be able to reconstruct the data into its original form.
  decode = function(s)
    if s == '' then return '' end
    local out = ''

    for match in s:gmatch('%d*.') do
      local last = match:len()
      local char = match:sub(last, last)
      local num = last == 1 and 1 or match:sub(1, last-1)
      out = out .. string.rep(char, num)
    end

    return out
  end
}

File: solutions/prolog/run-length-encoding/run_length_encoding.pl
%! encode(+Plaintext, -Ciphertext)
%
% The encode/2 predicate succeeds with Ciphertext as a string where 
% consecutive elements are represented as a count and data value.
%   For example: "AABBBCCCC" => "2A3B4C"
encode(Plaintext, Ciphertext) :-
    string_chars(Plaintext, Chars),
    clumped(Chars, Pairs),
    foldl(encode_pair, Pairs, [], Encoded),
    string_chars(Ciphertext, Encoded).

%! decode(+Ciphertext, -Plaintext)
%
% The decode/2 predicate succeeds with Plaintext as a string that 
% has been reconstructed from the cipher into its original form.
%   For example: "2A3B4C" => "AABBBCCCC"
decode(Ciphertext, Plaintext) :-
    re_split("[\\d]*[\\D]", Ciphertext, Splits),
    subtract(Splits, [""], Groups),
    foldl(decode_group, Groups, [], Decoded),
    string_chars(Plaintext, Decoded).

% encode_pair/3 converts a Pair into a CharList and appends it 
% to the Accumulator. For example: 'A'-12 => ['1', '2', 'A']
encode_pair(Letter-1, Acc, Encoded) :-
    append(Acc, [Letter], Encoded), !.
encode_pair(Letter-Count, Acc, Encoded) :-
    number_chars(Count, Chars),
    append([Acc, Chars, [Letter]], Encoded).

% decode_group/3 converts a String into a CharList and appends it
% to the Accumulator. For example: "3B" => ['B', 'B', 'B']
decode_group(Str, Acc, Decoded) :-
    string_chars(Str, Chars),
    last(Chars, Letter),
    append(Digits, [Letter], Chars),
    repeated(Letter, Digits, Letters),
    append(Acc, Letters, Decoded), !.

repeated(C, [], L) :- L = [C], !.
repeated(C, Ds, L) :-
    number_chars(N, Ds),
    length(L, N),
    maplist(=(C), L).